public/client/windows/TracyETW.cpp

```diff
@@ -0,0 +1,284 @@
+    tp.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED;
+    if( LookupPrivilegeValue( nullptr, PrivilegeName, &tp.Privileges[0].Luid ) == FALSE )
+        return ETWError( GetLastError() );
+    HANDLE hToken = {};
+    if( OpenProcessToken( GetCurrentProcess(), TOKEN_ADJUST_PRIVILEGES, &hToken ) == FALSE )
+        return ETWError( GetLastError() );
+    BOOL adjusted = AdjustTokenPrivileges( hToken, FALSE, &tp, 0, nullptr, nullptr );
+    DWORD status = GetLastError();
+    CloseHandle( hToken );    // NOTE: skipping error handling for CloseHandle()
+    return ETWError( status );
+}
+
+static ULONG StartSession( Session& session )
+{
+    ULONG status = StartTraceA( &session.handle, session.name, &session.properties );
+    if( status == ERROR_ALREADY_EXISTS )
+    {
+        // Session is already running (likely from a previous run that did not terminate
+        // gracefully). There are two options: take control of the existing session with
+        // ControlSession(UPDATE), or stop the session and start fresh again. The latter
+        // is better because it also resets the event providers.
+        status = ControlTraceA( session.handle, session.name, &session.properties, EVENT_TRACE_CONTROL_STOP );
+        if( status != ERROR_SUCCESS )
+            return ETWError( status );
+        status = StartTraceA( &session.handle, session.name, &session.properties );
+    }
+    return ETWError( status );
+}
+
+static ULONG StopSession( Session& session )
+{
+    ULONG status = ControlTraceA( session.handle, session.name, &session.properties, EVENT_TRACE_CONTROL_STOP );
+    if( status != ERROR_SUCCESS )
+        return ETWError( status );
+    // once stopped, the session handle becomes invalid
+    session.handle = 0;
+    return ERROR_SUCCESS;
+}
+
+static ULONG EnableProvider(
+    Session& session,
+    const GUID& ProviderId,
+    ULONG ControlCode = EVENT_CONTROL_CODE_ENABLE_PROVIDER,
+    UCHAR Level = TRACE_LEVEL_INFORMATION,
+    ULONGLONG MatchAnyKeyword = 0,  // NOTE: a MatchAnyKeyword of 0 actually means "all bits set", according to the EnableTraceEx2 docs
+    ULONGLONG MatchAllKeyword = 0,
+    ULONG Timeout = 0,
+    PENABLE_TRACE_PARAMETERS EnableParameters = NULL )
+{
+    ULONG status = EnableTraceEx2( session.handle, &ProviderId, ControlCode, Level, MatchAnyKeyword, MatchAllKeyword, Timeout, EnableParameters );
+    return ETWError( status );
+}
+
+static ULONG EnableStackWalk( Session& session, GUID EventGuid, UCHAR Opcode )
+{
+#if defined _WIN64
+    constexpr bool isOs64Bit = true;
+#else
+    BOOL _iswow64;
+    IsWow64Process( GetCurrentProcess(), &_iswow64 );
+    const bool isOs64Bit = _iswow64;
+#endif
+    if( !isOs64Bit )
+        return 0 /* ERROR_SUCCESS */;   // TODO: return error instead?
+    CLASSIC_EVENT_ID stackId[1] = {};
+    stackId[0].EventGuid = EventGuid;
+    stackId[0].Type = Opcode;
+    ULONG status = TraceSetInformation( session.handle, TraceStackTracingInfo, &stackId, sizeof( stackId ) );
```

Comment:
Is there any downside of doing multiple `TraceSetInformation` vs one call with an array of eventids as before?
I suppose the only difference might be performance / failure atomicity?

Reply:
If there is, it's not documented or noticeable in practice from what I can tell.